enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+  * Upstream [will offer some sort of dynamic provisioning in the next version](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/274)
+    by creating subdirectories (Access Points) in existing EFS volumes as PVs. AWS has a limit of 120 Access Points per
+    EFS volume, i.e. only 120 PVs can be created for a single StorageClass.
+    [Upstream issue](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/310).
+  * Proper dynamic provisioning can come in the future, dealing with the networking and security groups is the hardest
+    part there.
+* AWS EFS volumes do not have any size enforcement. They can hold as much data as apps put there and someone will have to
+  pay for the storage. OCP offer standard metrics for amount of data on PVCs, it's up to the cluster admin to create
+  alerts if the consumption gets too high.
+* It uses [efs-utils](https://github.com/aws/efs-utils) to provide encryption in transport, which is not included in
+  RHEL. We will need to package it, either as an image or RPM package.
+
+## Proposal
+High level only, see design details below.
+
+1. Rewrite current aws-efs-operator to use [CSI driver installation functions of library-go](https://github.com/openshift/library-go/tree/master/pkg/operator/csi/csicontrollerset).
+   * This allows us to share well tested code to install CSI drivers and implement bugfixes / new features at a single
+     place.
+   * It will add common features of CSI driver operators which are missing in the current community operator, such as
+     proxy injection and restarting the drivers when cloud credentials change.
+
+2. Update (rewrite) current aws-efs-operator to watch `ClusterCSIDriver` CR named `efs.csi.aws.com`.
+   * `ClusterCSIDriver` CRD is already provided by OCP and is used to install all other CSI drivers that OCP ships.
+   * Users must explicitly create the CR to install the CSI driver to allow users to uninstall the driver if they don't
+     like it.
+
+3. Keep existing functionality:
+   * `SharedVolume` CRD in `aws-efs.managed.openshift.io/v1alpha1` API group keeps working as it is now.
+   * Update from the community operator to the supported operator should be as seamless as possible.
+   * The operator is still installed via OLM.
+
+4. Ship the operator + CSI driver through ART pipeline.
+   * Most of the operator just installs / removes CSI driver, i.e. manages driver's RBAC rules, CredentialsRequest,
+     DaemonSet, Deployment, CSIDriver and StorageClasses. See existing
+     [AWS EBS operator](https://github.com/openshift/aws-ebs-csi-driver-operator/tree/master/assets) for objects it
+     manages, EFS should be very similar in this regard.
+   * `SharedVolume` processing on top of it is very trivial, it just creates PV + PVC.
+
+5. Ship [efs-utils](https://github.com/aws/efs-utils) as a base image for the EFS CSI driver image.
+
+6. Update installer to delete Access Points and EFS volumes that have `owner` tag when destroying a cluster.
```

Comment:
This feels wrong. We've talked about the barriers (networking, permissions, etc.) that would make this difficult for the operator to manage. IMO for those reasons -- and for symmetry -- all provisioning/deprovisioning on the AWS side should be outside of the purview of the operator.

(To be clear, I agree CI must have the ability to provision and deprovision -- more on this below -- but I don't think this tagging approach is the way to go.)

Reply:
Problem is that it's not possible to destroy a cluster if it has EFS volume. Installer will get stuck at destroying networks, because AWS does not allow to delete those used by EFS volumes. Since it's cluster admin who must create EFS share, it's up to them either to add `owner` tag (for automatic deletion on cluster teardown) or preserve it, but deal with the consequences.

Reply:
Added more detailed note about installer behavior and opt-in nature of the tag.

Reply:
I think there's still a disconnect here.

I'm not on board with the operator trying to talk to AWS. This will lead to confusion and complexity for negligible benefit.

It still requires the consumer (human or CI) to properly configure AWS access credentials. If they don't, the volume removal will fail, and they won't necessarily find out about it *except* in the case of a blocked cluster uninstall -- and even then, the cause may not be obvious.

It increases your test matrix substantially.

And it's asymmetrical. If the operator can handle deletion, it can easily handle provisioning as well. And that would make a lot more sense to the user.

I don't understand the objection to the CI test case/harness cleaning up explicitly.